lib/src/common/disposable.dart

```diff
@@ -26,19 +26,50 @@ abstract class _Disposable {
   Future<Null> dispose();
 }
 
-class InternalDisposable implements _Disposable {
+/// Used to invoke, and remove references to, a [Disposer] before disposal of the
+/// parent object.
+class ManagedDisposer implements _Disposable {
   Disposer _disposer;
+  final Completer<Null> _didDispose = new Completer<Null>();
+  bool _isDisposing = false;
 
-  InternalDisposable(this._disposer);
+  ManagedDisposer(Disposer disposer)
+      : _disposer = (disposer ?? () => new Future(() {}));
 
-  @override
+  /// A [Future] that will complete when this object has been disposed.
+  Future<Null> get didDispose => _didDispose.future;
+
+  /// Whether this object has been disposed.
+  bool get isDisposed => _didDispose.isCompleted;
+
+  /// Whether this object has been disposed or is disposing.
+  ///
+  /// This will become `true` as soon as the [dispose] method is called
+  /// and will remain `true` forever. This is intended as a convenience
+  /// and `object.isDisposedOrDisposing` will always be the same as
+  /// `object.isDisposed || object.isDisposing`.
+  bool get isDisposedOrDisposing => isDisposed || isDisposing;
+
+  /// Whether this object is in the process of being disposed.
+  ///
+  /// This will become `true` as soon as the [dispose] method is called
+  /// and will become `false` once the [didDispose] future completes.
+  bool get isDisposing => _isDisposing;
+
+  /// Dispose of the object, cleaning up to prevent memory leaks.
   Future<Null> dispose() {
-    var disposeFuture = _disposer != null ? _disposer() : null;
-    _disposer = null;
-    if (disposeFuture == null) {
-      return new Future.value();
+    if (isDisposedOrDisposing) {
+      return didDispose;
     }
-    return disposeFuture.then((_) => null);
+    _isDisposing = true;
+
+    var disposeFuture = _disposer() ?? new Future(() {});
```

Comment:
It's possible that the call to `_disposer` returns `null` rather than a `Future<Null>`, we actually have some tests that trigger that.